### PR TITLE
Add whitelist of trusted exchange rate oracles 

### PR DESCRIPTION
**Pull request by echevrier**
Update of exchange rate can only be done by a trusted oracle

### CODE REVIEW

File: teeracle/src/lib.rs [hunk 1]
```diff
@@ -51,31 +52,76 @@ pub mod pallet {
...
+			<Whitelist<T>>::try_append(mrenclave).map_err(|_| Error::<T>::TooManyOracles)?;
+			Self::deposit_event(Event::AddedToWhitelist(mrenclave));
+			Ok(())
+		}
+		#[pallet::weight(<T as Config>::WeightInfo::remove_from_whitelist())]
+		pub fn remove_from_whitelist(origin: OriginFor<T>, mrenclave: [u8; 32]) -> DispatchResult {
+			ensure_root(origin)?;
+			ensure!(Self::is_whitelisted(mrenclave), <Error<T>>::NonWhitelistedOracle);
+			Whitelist::<T>::mutate(|mrenclaves| {
+				mrenclaves.retain(|m| m.encode() != mrenclave.encode())
```
Comment (clangenb) [inline teeracle/src/lib.rs:111]: I guess you can spare the encode here.

File: teeracle/src/lib.rs [hunk 2]
```diff
@@ -97,6 +148,15 @@ pub mod pallet {
 		}
 	}
 }
+impl<T: Config> Pallet<T> {
+	fn is_whitelisted(mrenclave: [u8; 32]) -> bool {
+		Self::whitelist().iter().any(|m| m.encode() == mrenclave.encode())
+	}
```
Comment (clangenb) [inline teeracle/src/lib.rs:154]: If it is possible, you can also drop the encode here.

File: teeracle/src/lib.rs [hunk 3]
```diff
@@ -51,31 +52,76 @@ pub mod pallet {
...
-	pub(super) type ExchangeRates<T> =
+	pub(super) type ExchangeRates<T: Config> =
 		StorageMap<_, Blake2_128Concat, CurrencyString, ExchangeRate, ValueQuery>;
 
+	/// whitelist of trusted exchange rate oracles
+	#[pallet::storage]
+	#[pallet::getter(fn whitelist)]
+	pub(super) type Whitelist<T: Config> =
+		StorageValue<_, WeakBoundedVec<[u8; 32], T::MaxOracles>, ValueQuery>;
+
```
Comment (clangenb) [inline teeracle/src/lib.rs:71]: I have never heard about this, and couldn't find material. Can you give me some background about the weakbounded vec? :)

File: teeracle/src/tests.rs [hunk 4]
```diff
@@ -59,7 +74,8 @@ fn verifiy_update_exchange_rate_works() {
 fn verifiy_get_existing_exchange_rate_works() {
```
Comment (clangenb) [inline teeracle/src/tests.rs:74]: I just noticed some typos in the verify. :)

Comment (clangenb) [reply teeracle/src/tests.rs:74]: I am not fond of the verify prefix anyhow. I don't think it adds information, but only lengthens the function name unnecessarily.

**Review by clangenb — COMMENTED** [review]
Minor comments, but looks good.

I like the cleanly written tests!

File: teeracle/src/lib.rs (hunk 1 above)
Comment (echevrier) [reply teeracle/src/lib.rs:111]: Without encode: no implementation for `&[u8; 32] == [u8; 32]`

Comment (clangenb) [reply teeracle/src/lib.rs:111]: `|m| *m != mrenclave` should solve your problem then.

File: teeracle/src/lib.rs (hunk 3 above)
Comment (clangenb) [reply teeracle/src/lib.rs:71]: Found the info, thanks

Comment (echevrier) [reply teeracle/src/lib.rs:71]: See: https://github.com/paritytech/substrate/blob/master/frame/support/src/storage/weak_bounded_vec.rs
This is similar to BoundedVec, but less strict. It allows some methods to go beyond the vector boundary,  logging in such cases logs error messages, but have methods like try-append, try_mutate, which ensure not to go beyond some length
I used it to ensure the limit of the whitelist size.

File: teeracle/src/lib.rs (hunk 1 above)
Comment (echevrier) [reply teeracle/src/lib.rs:111]: Ouch

File: teeracle/src/tests.rs (hunk 4 above)
Comment (echevrier) [reply teeracle/src/tests.rs:74]: OK. I remove then all

**Review by clangenb — APPROVED** [review]
Looks good to me, nice and clean.

File: teeracle/src/lib.rs [hunk 5]
```diff
@@ -51,31 +51,74 @@ pub mod pallet {
...
 	}
 
 	#[pallet::error]
 	pub enum Error<T> {
 		InvalidCurrency,
+		/// Too many oracles in the whitelist.
+		TooManyOracles,
+		NonWhitelistedOracle,
+		AlreadyWhitelistedOracle,
+		UntrustedOracle,
```
Comment (brenzi) [inline teeracle/src/lib.rs:89]: This is the same as NonWhitelistedOracle
please merge the two errors. The origin of the error is different, but the meaning is the same IMO

File: teeracle/src/lib.rs [hunk 6]
```diff
@@ -51,31 +51,74 @@ pub mod pallet {
...
+		AddedToWhitelist([u8; 32]),
+		RemovedFromWhitelist([u8; 32]),
+		OracleWhitelistCleared,
 	}
 
 	#[pallet::error]
 	pub enum Error<T> {
 		InvalidCurrency,
+		/// Too many oracles in the whitelist.
+		TooManyOracles,
```
Comment (brenzi) [inline teeracle/src/lib.rs:86]: 1. You are not whitelisting Oracles (service instances), you are whitelisting Enclaves (binaries). I suggest you refactor `Oracle` to `MrEnclave` here and everywhere else where it makes sense
2. These are all "whitelisting" errors, so I think they should all have "Whitelist" in their name

Comment (brenzi) [reply teeracle/src/lib.rs:86]: suggestion:
```
MrEnclaveWhitelistOverflow
MrEnclaveNotWhitelisted
MrEnclaveAlreadyWhitelisted
```

File: teeracle/src/lib.rs [hunk 7]
```diff
@@ -97,6 +145,15 @@ pub mod pallet {
 		}
 	}
 }
+impl<T: Config> Pallet<T> {
+	fn is_whitelisted(mrenclave: [u8; 32]) -> bool {
+		Self::whitelist().iter().any(|m| *m == mrenclave)
```
Comment (brenzi) [inline teeracle/src/lib.rs:150]: why not `Self::whitelist().contains(mrenclave)` ? It's a vec, right?

File: teeracle/src/lib.rs [hunk 8]
```diff
@@ -97,6 +145,15 @@ pub mod pallet {
...
+	}
+
+	pub fn whitelisted_oracle_count() -> u32 {
```
Comment (brenzi) [inline teeracle/src/lib.rs:153]: why a helper function for this? the name is longer than the actual direct instruction

File: teeracle/src/mock.rs [hunk 9]
```diff
@@ -120,6 +120,7 @@ impl timestamp::Config for Test {
 parameter_types! {
 	pub const MomentsPerDay: u64 = 86_400_000; // [ms/d]
 	pub const MaxSilenceTime: u64 = 172_800_000; // 48h
+	pub const MaxOracles: u32 = 2;
```
Comment (brenzi) [inline teeracle/src/mock.rs:123]: rather `MaxWhitelistLength`?
and set it higher as 2 is very limiting for tests. maybe 10?

File: teeracle/src/lib.rs [hunk 10]
```diff
@@ -51,31 +51,74 @@ pub mod pallet {
...
+		pub fn remove_from_whitelist(origin: OriginFor<T>, mrenclave: [u8; 32]) -> DispatchResult {
+			ensure_root(origin)?;
+			ensure!(Self::is_whitelisted(mrenclave), <Error<T>>::NonWhitelistedOracle);
+			Whitelist::<T>::mutate(|mrenclaves| mrenclaves.retain(|m| *m != mrenclave));
+			Self::deposit_event(Event::RemovedFromWhitelist(mrenclave));
+			Ok(())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::clear_whitelist())]
+		pub fn clear_whitelist(origin: OriginFor<T>) -> DispatchResult {
```
Comment (brenzi) [inline teeracle/src/lib.rs:115]: what is this function good for? I don't see a reason to have it as we can remove all until empty one by one. It's just one more function to maintain

**Review by brenzi — CHANGES_REQUESTED** [review]

File: teeracle/src/lib.rs (hunk 6 above)
Comment (echevrier) [reply teeracle/src/lib.rs:86]: I think these errors should be user-friendly. And I'm not sure that `MrEnclaveNotWhitelisted`, ... is clear for a dummy user.
From a user's perspective, is it wrong to say it is about the whitelisting of oracles?  In the issue description, you wrote "_the teeracle should maintain a whitelist of oracles_".

File: teeracle/src/mock.rs (hunk 9 above)
Comment (echevrier) [reply teeracle/src/mock.rs:123]: This is a param that has to be configured in the Node runtime. Here again, I think that MaxWhitelistLength is not so "user-friendly". If "oracle" is ok, then what about MaxTrustedOracles ?

File: teeracle/src/lib.rs (hunk 6 above)
Comment (brenzi) [reply teeracle/src/lib.rs:86]: It seems I wasn't precise. What we are whitelisting is MrEnclave. But it is true that MrEnclave is very SGX specific and not generic for all TEE technologies.
What about `Release`? Each oracle service runs a specific release. And MrEnclave identifies a release unambiguously

Comment (echevrier) [reply teeracle/src/lib.rs:86]: OK

File: teeracle/src/mock.rs (hunk 9 above)
Comment (echevrier) [reply teeracle/src/mock.rs:123]: What about this?

Comment (brenzi) [reply teeracle/src/mock.rs:123]: MaxWhitelistedReleases

**Review by brenzi — APPROVED** [review]

**Pull request merged**